library/bignum.c

```diff
@@ -2205,23 +2215,23 @@ int mbedtls_mpi_gen_prime( mbedtls_mpi *X, size_t nbits, int dh_flag,
 
     n = BITS_TO_LIMBS( nbits );
 
+new_number:
     MBEDTLS_MPI_CHK( mbedtls_mpi_fill_random( X, n * ciL, f_rng, p_rng ) );
+    /* make sure generated number is at least (nbits-1)+0.5 bits (FIPS 186-4 §B.3.3 steps 4.4, 5.5) */
+    if( X->p[n-1] < CEIL_MAXUINT_DIV_SQRT2 ) goto new_number;
```

Comment:
Note: This breaks the MPI abstraction layer. One way to make respect it would be to setup an MPI from `CEIL_MAXUINT_DIV_SQRT2`, shift it appropriately, and call `mbedtls_mpi_cmp`. @mpg WDYT?

Reply:
OTOH, the pre-existing https://github.com/ARMmbed/mbedtls/pull/1380/files#diff-c252711ea8b2703e393502bfe46bff62R2245 below uses the internal structure of `mbedtls_mpi`, too.

Reply:
I think this is inside `bignum.c`, not `rsa.c`, so it's OK to break the MPI abstraction :)

Reply:
Omg... sorry!